File: python/src/models/event.py
class Event:
    def __init__(self, type: str):
        self.type = type

    def toJson(self) -> dict:
        return {"type": self.type}

    def fromJson(self, json: dict):
        self.type = str(json.get("type"))

    @staticmethod
    def parse(json: dict) -> "Event":
        from .custom_event import CustomEvent
        from .keyboard_event import KeyboardEvent
        from .mouse_event import MouseEvent

        type = str(json.get("type"))
        switch = {
            "mouse": MouseEvent,
            "keyboard": KeyboardEvent,
            "custom": CustomEvent,
        }
        return switch[type].fromJson(json)

    def __str__(self):
        from .custom_event import CustomEvent
        from .keyboard_event import KeyboardEvent
        from .mouse_event import MouseEvent

        if type(self) is MouseEvent:
            return str(
                MouseEvent(
                    self.mouseEventType, self.x, self.y, self.button, self.dx, self.dy
                )
            )
        elif type(self) is KeyboardEvent:
            return str(KeyboardEvent(self.state, self.key, self.specialKey))
        elif type(self) is CustomEvent:
            return str(CustomEvent(self.customCommand, self.delay))
        else:
            return f"Unknown event type: {self}"

File: lib/models/events/event.dart
import 'custom/custom_event.dart';
import 'keyboard/keyboard_event.dart';
import 'mouse/mouse_event.dart';

abstract class Event {
  final String type;

  Event({required this.type});

  Map<String, dynamic> toJson() {
    return {
      'type': type,
    };
  }

  Event.fromJson(Map<String, dynamic> json) : type = json['type'].toString();

  Event clone();

  static Event parse(Map<String, dynamic> json) {
    final type = json['type'].toString();
    switch (type) {
      case KeyboardEvent.eventType:
        return KeyboardEvent.fromJson(json);
      case MouseEvent.eventType:
        return MouseEvent.fromJson(json);
      case CustomEvent.eventType:
        return CustomEvent.fromJson(json);
      default:
        throw Exception('Unknown event type: $type');
    }
  }
}
